### DISCUSSION TITLE
Puzzle Generation in Chess Forge

### DISCUSSION THREAD

**Discussion by akaKarthick** [czbar/ChessForge#768] [Ideas]
I am writing to suggest a new feature for Chess Forge: the ability to **generate random positions** for puzzles with hidden solutions. I believe that this feature would be a valuable addition to the program, making it more engaging, educational, and accessible to users of all skill levels.

- **Endless variety:** This feature would provide users with an **endless supply of new puzzles** to solve, helping them to stay motivated and engaged. (I tried to import puzzles from the Lichess database, but it was too large. I broke it down into smaller PGN files, each containing 5,000 puzzles, but the program still couldn't handle it. So being able to generate puzzles based on user preferences would be very helpful.)

- Customizable difficulty: Users would be able to customize the difficulty of the puzzles by choosing the following parameters:

1. **Eval bar diff:** This parameter would control the complexity of the puzzle, with higher eval bar differences indicating more complex puzzles.

2. **Number of pieces:** This parameter would control the number of pieces on the board, with higher numbers of pieces indicating more complex puzzles.

3. **Forced mate:** This **checkbox** would allow users to generate puzzles with forced mates.

4. **Material balance:** This parameter would control the material balance of the position, with higher values indicating more balanced positions.

5. **Piece types:** This parameter would allow users to specify the types of pieces that they want to include in the puzzle. For example, a user could choose to generate a puzzle with only knights and pawns.

This level of customization would make the feature accessible to players of all skill levels, and it would also allow users to focus on **specific areas** of their game that **they need to improve**. For example, a user who is **struggling with endgames** could choose to generate a puzzle with a **forced mate**.

What are your thoughts on this feature suggestion?
Do you think it is something that is possible to implement?
Are there any other features that you are currently planning to implement in Chess Forge?
I look forward to hearing from you soon.

**Comment by czbar** [comment]
Before I comment on the main subject:

**A quick question**: you mentioned your attempt to import studies via PGN files with 5,000 exercise each. ChessForge should be able to import such numbers quite easily so I presume there was something more to that scenario.  If you can send me details, I would look into it.

**Chess Forge new features and future**: with version 1.4 (published yesterday) I have implemented all the **core** functionality that I have planned for Chess Forge. In particular, the new "Online Library" feature was designed to attract potential users who want to just consume content and are not interested in authoring stuff. 

I will keep working on the program, but I will be dedicating much less time to it. So, small features, bug fixes and improvements but nothing big (e.g. no "spaced repetition") ... unless the program becomes really popular.  Last month an article was published in a leading Polish chess magazine about Chess Forge and it drove downloads to over 100 a week but it has quickly subsided to just a few a day. And yesterday there was a bump (25 downloads) after I published a new version.  Later today or tomorrow I will publish a short blog on lichess regarding the "Online Library" and will see if it has any effect.  Lichess blogs don't get many views (my first one from April had just 300 views) and also lichess has some secret algorithm that puts some blogs on the main blog page but not others. Mine was never put there so people had to dig deeper to find it.  

Obviously, a mention from some youtube chess celebrity would make a huge difference but I don't know personally any of them.

In summary, I'll watch the popularity of the program and will invest my time in its development accordingly.

**Comments on your proposal**: this is super interesting but also super difficult. I am not sure if anyone attempted this kind of work before. I may check with my contacts in the Chess Composition community if they have tried something of the kind.  They have different requirements and use different software than "regular" chess players so maybe something like that was thought of before.

Why did I immediately think of chess composition?   Because a random generation would likely result in positions that look more like those artificial compositions than positions from real games!  

Generating random, legal positions would not be a problem but selecting those that, kind of, make sense, would be.  I believe, it is absolutely possible, though, as we can probably define criteria that make positions look "sensible".

Then, there is a problem of complexity/difficulty.  I think, you are using these two terms interchangeably, but I would define them as two different attributes.  I'd say that _complexity_ would relate to the number of pieces and possibilities on the board, while _difficulty_ would reflect how hard it is to solve the problem.  Those two characteristics could go hand-in-hand but could be very different in different positions.

In my definition above, _complexity_ would be easy to measure (just need some strict definition) while _difficulty_ would be extremely hard to evaluate.  You can see for yourself how badly rated are puzzles on chess.com and lichess.  And they don't even attempt to automate it, just relying on how humans with different ratings are successful in solving them.  Also, those sites are embarrassingly bad at detecting "great" moves. Chess.com often gives me "!!" for an almost forced move.  The only reliable evaluation is blunder detection which is why I implemented it in Chess Forge :)

Nevertheless, I like the idea, it has a potential to be quite revolutionary.  I'll explore it a bit outside of Chess Forge. As I mentioned, I'll educate myself a little on Chess Composition which is an area that I was never really much into.

**Reply by akaKarthick** [reply]
https://x.com/lichess/status/1735334542959939659?s=20
